web/src/app/components/timeline/timeline.component.ts

```diff
@@ -0,0 +1,434 @@
+import { AfterViewInit, Component, ElementRef, Input, ViewChild } from '@angular/core';
+import { HumanizeDurationLanguage, HumanizeDuration, HumanizeDurationOptions } from 'humanize-duration-ts';
+import * as d3 from 'd3';
+
+import { step189_2020 } from '../../../proto/step189_2020';
+import { ScaleTime, min } from 'd3';
+
+/**
+ * Item holds all required data for one interval on the timeline.
+ */
+interface Item {
+  pushID: string;     // Push ID number
+  state: number;      // Final state of push
+  startTime: number;  // Start of push, in milliseconds
+  endTime: number;    // End of push, in milliseconds
+  row: number;        // Row number corresponds to y-position on timeine
+}
+
+@Component({
+  selector: 'app-timeline',
+  templateUrl: './timeline.component.html',
+  styleUrls: ['./timeline.component.scss']
+})
+
+export class TimelineComponent implements AfterViewInit {
+  /**
+   * Private data variables. 
+   * 
+   * Note that we use the non-null assertion operator ('!') to indicate to 
+   * the compiler that variables will always be declared and thus to not issue
+   * errors about possibilities of them being `null` or `undefined` since
+   * pushInfos and data are passed-in values.
+   */
+  @ViewChild('timeline') private timelineContainer!: ElementRef;
+  @Input() private pushInfos!: step189_2020.IPushInfo[] | null;
+  private data: Item[] = [];
+  private svg: any;
+  private x: ScaleTime<number, number> = d3.scaleTime();
+  private xAxis: Function = () => { };
+  private height: number = 0;
+  private width: number = 0;
+  private numRows: number = 0;
+
+  /**
+   * Constants.
+   */
+  private static readonly LANG_SERVICE: HumanizeDurationLanguage = new HumanizeDurationLanguage();
+  private static readonly HUMANIZER: HumanizeDuration = new HumanizeDuration(TimelineComponent.LANG_SERVICE);
+  private static readonly COLOR_LIGHT_GRAY: string = '#d3d3d3';
+  private static readonly MIN_INTERVAL_HEIGHT: number = 25;
+  private static readonly MSEC_PER_MIN: number = 6 * (10 ** 4);
```

Comment:
```suggestion
  private static readonly MSEC_PER_MIN: number = 60 * (10 ** 3);
```
60 * (10 ** 3) is more intuitive to me since it's msec to sec and then to minute.